File: app/models/Domain.java
package models;

import java.util.Collection;

import models.questions.Question;

public class Domain {
	public static class QuestionAnswerResponse {
		public boolean correct;
		public int score;
		public String type = "answerResponse";
		
		public QuestionAnswerResponse(boolean correct, int score) {
			this.correct = correct;
			this.score = score;
		}
	}
	
	public static class RegistrationResponse {
		public String type = "registrationResponse";
		public int statusCode = 200;
		public Team team;
		
		public RegistrationResponse(Team team) {
			this.team = team;
		}
		
	}
	
	public static class ErrorResponse {
		
		public String type = "error";
		public String message;
		
		public ErrorResponse(String message) {
			this.message = message;
		}
	}
	
	public static class TeamListResponse {
		public String type = "teamList";
		public Collection<Team> teams;
		
		public TeamListResponse(Collection<Team> teams) {
			this.teams = teams;
		}
	}
	
	public static class QuestionAdminResponse {
		public String type = "currentQuestion";
		public Question question;
		
		public QuestionAdminResponse(Question question) {
			this.question = question;
		}
	}
	
	public static class BuzzAck {
		public String type = "buzzAck";
		public int responseOrder;
		public String teamName;
		
		public BuzzAck(String teamName, int responseOrder) {
			this.teamName = teamName;
			this.responseOrder = responseOrder;
		}
	}
	
	public static class Scored {
		public String type = "scored";
		public int score;
		
		public Scored(int score) {
			this.score = score;
		}
	}
}

File: appKotlin/models/Domain.kt
package models

sealed class Domain(val type: String) {

    class QuestionAnswerResponse(var correct: Boolean, var score: Int): Domain("answerResponse")

    class RegistrationResponse(var team: Team): Domain("registrationResponse") {
        val statusCode = 200
    }

    class ErrorResponse(var message: String): Domain("error")

    class TeamListResponse(var teams: Collection<Team>): Domain("teamList")

    class QuestionAdminResponse(var question: Question): Domain("currentQuestion")

    class BuzzAck(var teamName: String, var responseOrder: Int): Domain("buzzAck")

    class Scored(var score: Int): Domain("scored")

    abstract class Question(val answerType: AnswerType, val questionNumber: Int, val question: String): Domain("question") {
        enum class AnswerType {
            SIMPLE, BUZZER
        }

        abstract fun checkAnswer(answer: String): Boolean
    }

    class SimpleQuestion(questionNumber: Int, question: String, private val answer: String) : Question(Question.AnswerType.SIMPLE, questionNumber, question) {

        override fun checkAnswer(answer: String): Boolean {
            play.Logger.debug("checking answer " + answer + " against " + this.answer)
            return answer.equals(this.answer, ignoreCase = true)
        }
    }

    class BuzzerQuestion(questionNumber: Int, question: String) : Question(AnswerType.BUZZER, questionNumber, question) {

        override fun checkAnswer(answer: String): Boolean {
            return false
        }
    }


}